packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -163,6 +176,20 @@ export default createRule<Options, MessageId>({
         node.test !== null &&
         node.test.type !== AST_NODE_TYPES.LogicalExpression
       ) {
+        /**
+         * Allow:
+         *   while (true) {}
+         *   for (;true;) {}
+         *   do {} while (true)
+         */
+        if (
+          allowConstantLoopConditions &&
+          isLoopStatement(node) &&
```

Comment:
Nit: I think we could refactor a bit to avoid this check.  

Instead of registering handlers like:

```ts
ConditionalExpression: checkIfTestExpressionIsNecessaryConditional,
DoWhileStatement: checkIfTestExpressionIsNecessaryConditional,
ForStatement: checkIfTestExpressionIsNecessaryConditional,
IfStatement: checkIfTestExpressionIsNecessaryConditional,
WhileStatement: checkIfTestExpressionIsNecessaryConditional,
```

we could change it to:

```
ConditionalExpression: checkIfTestExpressionIsNecessaryConditional,
DoWhileStatement: checkIfLoopIsNecessaryConditional,
ForStatement: checkIfLoopIsNecessaryConditional,
IfStatement: checkIfTestExpressionIsNecessaryConditional,
WhileStatement: checkIfLoopIsNecessaryConditional,
```

And include the `allowConstantLoopConditions` logic only inside `checkIfLoopIsNecessaryConditional`.  This would leverage the AST listener a little better and avoid some unnecessary checks.  (And hopefully improve performance of this rule a little bit)

Reply:
I think this is probably worthwhile, +1